File: Week 6/Java/src/com/company/Raygun.java
/*
* Classname: Raygun.java
*
* Version information: 1.0
*
* Date: 21/01/2021
*
* Description: The class implements a ray gun for a game
*
* Author: Dr. Shane Wilson
*
*/

package com.company;

public class Raygun {
    //Data members
    private int mChrg; //The maximum charge the Raygun can hold
    private int cChrg; //The current level of charge the Raygun has

    //Class constructors
    public Raygun(int mChrg, int cChrg) {
        //TODO: Construct the Raygun object and initialise the attributes
    }

    // Methods
    //TODO: Implement appropriate getters and setters for the class

    public void fireAt(Alien a) {
        //TODO: Implement the method
        /*The gun should only fire if it is charged. Firing the gun will reduce the charge by 1. If the alien isDodging
        * alien.miss() should be called, otherwise alien.hit */

    }

    public void recharge() {
        //TODO: Implement the method
        /*Increases the current charge by 1*/
    }

    public boolean isCharged() {
        //TODO: Implement the method
        /*Returns true if the gun has a charge*/

    }

    public boolean isFullyCharged() {
        //TODO: Implement the method
        /*Returns true if the gun is fully charged*/
    }
}

File: Week 6/C Sharp/Aliens/Aliens/Models/Raygun.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using static System.Runtime.InteropServices.JavaScript.JSType;

namespace Aliens.Models
{
    internal class Raygun
    {
        // Properties
        
        //The maximum charge the Raygun can hold
        public int MaxCharge { get; set; }

        //The current level of charge the Raygun has
        public int CurrentCharge { get; set; }

        public Raygun(int mChrg, int cChrg)
        {
            MaxCharge = mChrg;
            CurrentCharge = Math.Clamp(cChrg, 0, MaxCharge);
        }

        // Methods
        

        public void FireAt(Alien a)
        {
            {
                if (IsCharged()) // Check if the gun has charge
                {
                    CurrentCharge--; // reduce by 1

                    if (a.Dodging)
                    {
                        a.Miss(); // dodged
                    }
                    else
                    {
                        a.Hit(); //hit
                    }
                }
            }
            /*The gun should only fire if it is charged. Firing the gun will reduce the charge by 1. If the alien isDodging
            * alien.miss() should be called, otherwise alien.hit */

        }

        public void Recharge()
        {
            if (!IsFullyCharged())
            {
                CurrentCharge++;  //increase by1
            }
        }

        public bool IsCharged()
        {
            return CurrentCharge > 0;

        }

        public bool IsFullyCharged()
        {
            return CurrentCharge == MaxCharge;
        }


    }
}
